File: tools/govectors/multiport.go
package main

// Area "multiport" (plan step 08.4): the two address parsers kcptun uses before it opens a
// socket.
//
//   - std.ParseMultiPort (kcptun std/multiport.go, copied verbatim into internal/std) turns
//     "host:port" or "host:minport-maxport" into a *MultiPort. It is regexp-based
//     (`(.*)\:([0-9]{1,5})-?([0-9]{1,5})?`, unanchored and leftmost-first), so it accepts
//     surrounding junk and produces surprising splits for out-of-range ports; the cases below
//     pin every one of those quirks down.
//   - net.SplitHostPort (Go standard library, net/ipsock.go) is used for its *failure*:
//     client/main.go:319 and server/main.go:449 treat an address it rejects as the path of a
//     unix socket, so which addresses it rejects (and with which message) decides whether
//     kcptun listens on TCP or on AF_UNIX.
//
// Each case records the function, the address, whether it succeeded, the results and the
// error text (errors.Errorf / *net.AddrError, both printed with %v as kcptun's log.Println
// prints them). Cases are grouped by function; within a group the order is the order of the
// address table below.

import (
	"net"

	"github.com/kcptun-rust/tools/govectors/internal/std"
)

// multiportCase is one std.ParseMultiPort vector.
type multiportCase struct {
	Name string `json:"name"`
	Func string `json:"func"` // always "ParseMultiPort"
	Addr string `json:"addr"`
	OK   bool   `json:"ok"`
	// Host, MinPort and MaxPort are the *MultiPort fields; only present on success.
	Host    string `json:"host,omitempty"`
	MinPort uint64 `json:"minport,omitempty"`
	MaxPort uint64 `json:"maxport,omitempty"`
	// Err is the error text ("" on success).
	Err string `json:"err,omitempty"`
}

func (c multiportCase) CaseName() string { return c.Name }

// splitHostPortCase is one net.SplitHostPort vector.
type splitHostPortCase struct {
	Name string `json:"name"`
	Func string `json:"func"` // always "SplitHostPort"
	Addr string `json:"addr"`
	OK   bool   `json:"ok"`
	Host string `json:"host,omitempty"`
	Port string `json:"port,omitempty"`
	Err  string `json:"err,omitempty"`
}

func (c splitHostPortCase) CaseName() string { return c.Name }

// multiportAddrs are the addresses ParseMultiPort is run on. The comment on each line is the
// behaviour it pins down; the plan step names the first nine.
var multiportAddrs = []string{
	":29900",                 // no host, the kcptun server's default -listen
	"1.2.3.4:3000-4000",      // a port range
	"[::1]:1-2",              // a bracketed IPv6 literal keeps its brackets in Host
	"vps:29900",              // the client's default -remoteaddr
	"host:123456",            // 6 digits: 12345 + 6, then the range check fails
	"a:0",                    // port 0 is rejected
	"a:5-3",                  // minport > maxport
	"nocolon",                // no match at all
	"x:65536",                // above 65535
	"",                       // empty address
	":",                      // a colon with no digits: no match
	"a:",                     // ditto
	"a:1",                    // the smallest accepted port
	"a:65535",                // the largest accepted port
	"a:1-65535",              // the widest accepted range
	"a:5-5",                  // a single-port range
	"a:012",                  // strconv.Atoi is base 10: 12, not octal 10
	"a:00",                   // "00" parses as 0 and is rejected
	"a:0-1",                  // a zero endpoint is rejected even inside a range
	"a:1-0",                  // ditto, on the other side
	"a:1-",                   // a dangling '-' is simply not part of the match
	"a:1-2-3",                // only the first two numbers are looked at
	"a:1-2junk",              // unanchored: trailing junk is ignored
	"junk a:1-2",             // unanchored: leading junk lands in Host
	"host:80:90",             // greedy (.*): the LAST colon wins, so Host is "host:80"
	"[2001:db8::1]:443",      // a real IPv6 address
	"2001:db8::1:443",        // without brackets the last colon still wins
	"unix:/tmp/kcptun.sock",  // a unix socket path: no digits after the colon, no match
	"/tmp/kcptun.sock",       // ditto without a colon
	"sub.domain.example:443", // a normal hostname
	"HOST:80",                // the regexp is case-blind; Host is copied verbatim
	"héllo:80",               // non-ASCII host, byte-exact in Host
	"a\nb:80",                // '.' does not match '\n', so the match starts after it
	"a:80\n",                 // a trailing newline is outside the match
	"a:1234567",              // 7 digits: 12345 + 67
	"a:99999",                // 5 digits, above 65535
	"-5:80",                  // a '-' in the host is nothing special
	"a:-1",                   // the sign is not part of ([0-9]{1,5})
	"::80",                   // Host is ":"
	"a:1e2",                  // only the leading digits match
}

// splitHostPortAddrs are the addresses net.SplitHostPort is run on: kcptun's own -l / -t
// shapes plus every branch of net/ipsock.go:SplitHostPort.
var splitHostPortAddrs = []string{
	":29900",                    // the server's default -listen: an empty host
	"127.0.0.1:12948",           // the client's default -localaddr shape
	"vps:29900",                 // a hostname
	"1.2.3.4:3000-4000",         // a port range: SplitHostPort does not validate the port
	"[::1]:80",                  // a bracketed IPv6 literal, brackets removed
	"[fe80::1%eth0]:80",         // a zone identifier stays in Host
	"[2001:db8::1]:443",         //
	"host:",                     // an empty port is accepted
	"host:port",                 // the port is not parsed
	"nocolon",                   // "missing port in address"
	"",                          // the empty address: the error carries no address
	":",                         // host "" port ""
	"::1:80",                    // unbracketed IPv6: "too many colons in address"
	"[::1]",                     // "missing port in address"
	"[::1]80",                   // ']' not followed by ':': "missing port in address"
	"[::1]:80:90",               // ']' followed by a non-final colon: "too many colons"
	"[::1:80",                   // "missing ']' in address"
	"a]:80",                     // "unexpected ']' in address"
	"a[b:80",                    // "unexpected '[' in address"
	"[a]b:80",                   // ']' is followed by 'b': "missing port in address"
	"/tmp/kcptun.sock",          // a unix path: rejected, which is how kcptun detects one
	"/tmp/kcptun.sock:80",       // a path with a port-looking suffix is accepted
	"unix:/tmp/kcptun.sock",     // accepted: host "unix", port "/tmp/kcptun.sock"
	"./relative.sock",           // rejected
	"@abstract",                 // rejected (Linux abstract socket name)
	"[]:80",                     // an empty bracketed host is accepted
	"[]",                        // "missing port in address"
	"héllo:80",                  // non-ASCII host
	"host:80\n",                 // the port is taken verbatim, newline included
	"[::ffff:1.2.3.4]:80",       // an IPv4-mapped IPv6 literal
	"::ffff:1.2.3.4:80",         // the same without brackets: too many colons
	"kcptun-server.example:443", //
}

// genMultiport runs both parsers over their address tables.
func genMultiport() ([]any, error) {
	cases := make([]any, 0, len(multiportAddrs)+len(splitHostPortAddrs))
	for _, addr := range multiportAddrs {
		c := multiportCase{Name: "parse/" + addr, Func: "ParseMultiPort", Addr: addr}
		mp, err := std.ParseMultiPort(addr)
		if err != nil {
			c.Err = err.Error()
		} else {
			c.OK = true
			c.Host, c.MinPort, c.MaxPort = mp.Host, mp.MinPort, mp.MaxPort
		}
		cases = append(cases, c)
	}
	for _, addr := range splitHostPortAddrs {
		c := splitHostPortCase{Name: "splithostport/" + addr, Func: "SplitHostPort", Addr: addr}
		host, port, err := net.SplitHostPort(addr)
		if err != nil {
			c.Err = err.Error()
		} else {
			c.OK = true
			c.Host, c.Port = host, port
		}
		cases = append(cases, c)
	}
	return cases, nil
}

File: crates/std/src/multiport.rs
//! Port of kcptun's multi-port address parser (`reference/kcptun/std/multiport.go`).
//!
//! The client's `-remoteaddr` and the server's `-listen` may name a range of UDP ports
//! (`IP:minport-maxport`), over which kcptun spreads its connections. Go parses them with one
//! unanchored regular expression and a hand-written range check, and the port reproduces both,
//! quirks included:
//!
//! - the pattern is `(.*)\:([0-9]{1,5})-?([0-9]{1,5})?`, applied with
//!   `FindStringSubmatch`, so it is **unanchored** and matches leftmost-first with a greedy
//!   `.*`: `host:80:90` has the host `host:80`, `junk a:1-2` the host `junk a`, and trailing
//!   text after the ports is ignored;
//! - at most five digits are taken per port, so `host:123456` parses as minport 12345 and
//!   maxport 6 — and is then rejected by the range check;
//! - `-?` is *outside* the third group, so `a:1-` is a valid single port.
//!
//! Go's `regexp` and the Rust `regex` crate implement the same leftmost-first semantics on the
//! same pattern, and `.` matches any character except `\n` in both, so the same sub-matches
//! come out; `testdata/vectors/multiport.json` proves it case by case against the Go code.

use std::sync::LazyLock;

use regex::Regex;

/// The address pattern, compiled once.
///
/// Note the escaped `\:`, which is redundant in both engines but kept verbatim so the pattern
/// can be compared with Go's at a glance.
// Go: kcptun/std/multiport.go:remoteAddrMatcher
static REMOTE_ADDR_MATCHER: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"(.*)\:([0-9]{1,5})-?([0-9]{1,5})?").expect("the address pattern is valid")
});

/// A parsed multi-port address: the host and an inclusive range of ports.
// Go: kcptun/std/multiport.go:MultiPort
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MultiPort {
    /// Everything before the last colon of the match, verbatim (brackets included).
    pub host: String,
    /// The first port of the range.
    pub min_port: u64,
    /// The last port of the range; equal to `min_port` for a single port.
    pub max_port: u64,
}

/// Why an address is not a usable multi-port address. The texts are Go's
/// `errors.Errorf` messages; kcptun prints them with `log.Println`, which shows the message
/// alone (the `github.com/pkg/errors` stack trace only appears under `%+v`).
// Go: kcptun/std/multiport.go:ParseMultiPort
#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
pub enum MultiPortError {
    /// The ports are out of order, zero, or above 65535.
    #[error("invalid port range specified: minport:{min_port} -> maxport {max_port}")]
    InvalidPortRange {
        /// The first port, as parsed (may be out of range).
        min_port: i64,
        /// The last port, as parsed (may be out of range).
        max_port: i64,
    },
    /// The address does not contain `:` followed by digits at all.
    #[error("malformed address:{addr}")]
    MalformedAddress {
        /// The address as it was given.
        addr: String,
    },
}

/// Parses a multiport listener or dialer address.
// Go: kcptun/std/multiport.go:ParseMultiPort
pub fn parse(addr: &str) -> Result<MultiPort, MultiPortError> {
    // Go: matches := remoteAddrMatcher.FindStringSubmatch(addr); if len(matches) >= 4 — the
    // pattern has three groups, so a match always has four entries and a miss none.
    if let Some(matches) = REMOTE_ADDR_MATCHER.captures(addr) {
        // Go: strconv.Atoi. Both groups are one to five decimal digits, so neither call can
        // fail; Go's `err` branches are unreachable and have no counterpart here.
        let min_port = atoi(matches.get(2).map_or("", |m| m.as_str()));
        let mut max_port = min_port;

        // multiport assignment
        let third = matches.get(3).map_or("", |m| m.as_str());
        if !third.is_empty() {
            max_port = atoi(third);
        }

        if (min_port > max_port)
            || min_port > 65535
            || max_port > 65535
            || min_port == 0
            || max_port == 0
        {
            return Err(MultiPortError::InvalidPortRange { min_port, max_port });
        }

        return Ok(MultiPort {
            host: matches.get(1).map_or("", |m| m.as_str()).to_string(),
            // Go: uint64(minPort) / uint64(maxPort); both are in 1..=65535 here.
            min_port: min_port as u64,
            max_port: max_port as u64,
        });
    }

    Err(MultiPortError::MalformedAddress {
        addr: addr.to_string(),
    })
}

/// `strconv.Atoi` for the one-to-five-digit groups of the pattern; Go's `int` is 64 bits.
fn atoi(digits: &str) -> i64 {
    digits
        .parse::<i64>()
        .expect("the pattern matched one to five decimal digits")
}

#[cfg(test)]
#[path = "multiport_tests.rs"]
mod tests;
